doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,156 @@
+# Read Holds
+
+- Durable Subscribes epic: [17541](https://github.com/MaterializeInc/materialize/issues/17541)
+- Configurable compaction windows: [16701](https://github.com/MaterializeInc/materialize/issues/16701)
+
+## The Problem
+
+Users are not able to use `SUBSCRIBE` without data loss.
+When materialize restarts or a client connection fails for some reason,
+users are not able to restart the `SUBSCRIBE` `AS OF` their most recently observed timestamp because the read frontier has usually progressed past that time.
+This happens because the compaction window is 1 second for all user objects.
+
+## Success Criteria
+
+`SUBSCRIBE` must be usable without data loss across server restarts or client reconnections.
+
+## Solution Proposal
+
+Add a new `HOLD` object that acquires controller read holds on listed objects at a specific timestamp.
+The `HOLD` must be manually advanced to newer timestamps by users which will release the previous and acquire a new read hold to the specified timestamp.
+If the timestamp falls behind a configurable window, the `HOLD` is marked in an error state and the read holds are released.
+
+The `HOLD` can be created by:
+
+```
+CREATE HOLD <name>
+ON <object1>[, <object2> ...]
+[AT <time>]
+[WITH (
+    MAX LAG = '3h'::INTERVAL,
+)]
+```
+
+A `HOLD` can be created on indexes, materialized views, tables, and sources.
+It cannot exist on sinks or unmaterialized views.
+If a `HOLD` exists for a directly queryable object (i.e., not indexes) at time `t`, `SELECT` and `SUBSCRIBE` on that object `AS OF t` will successfully execute.
+If a `HOLD` exists for an index, the same is true if the index satisfies the query.
+A `HOLD` on any non-index object will not also create the `HOLD` on any indexes of the object.
+An `AS OF` query will restrict the indexes or object it considers to those that are valid for the requested time.
+
+Because this may be surprising to users, if an `AS OF` query whose plan uses an index errors because the `AS OF` is less than the index's read frontier and a `HOLD` exists on the index's object,
+a helpful error message will be produced telling the user that their `HOLD` must be on the index, not the underlying object.
+
+The objects the `HOLD` is created on do not need to be in the same schema or database.
+The `HOLD` can be created in any schema on which the user has a `CREATE` privilege (see (Alternatives)[#Alternatives] for other takes on this).
+The user must have a `SELECT` privilege on all objects.
```

Comment:
If we wanted to treat it like `ALTER`, then we'd require ownership of an object to install a read hold, which is probably too restrictive. I like the idea of adding `USAGE` on any clusters that may be effected since we'll affect the memory footprint of those clusters. However, that doesn't address read holds on tables or sources. Those still only require `SELECT` which feels a little off.

What about introducing a custom privilege type called `HOLD`, that is required on an object in order to install a read hold? Or maybe a system privilege that allows a user to install a read hold on any object.

Reply:
Indeed, I like the comparison to `SELECT`s! But—`SELECT`s do require `USAGE` on the cluster on which they run, right?

Reply:
> But—`SELECT`s do require `USAGE` on the cluster on which they run, right?

Yes that's correct (https://materialize.com/docs/sql/select/#privileges).

Reply:
Added USAGE.